decryptor/mysql/data_encoder.go

```diff
@@ -0,0 +1,388 @@
+type DataEncoderProcessor struct {
+	BaseMySQLDataProcessor
+}
+
+// NewDataEncoderProcessor return new data encoder from/to binary format for tokenization
+func NewDataEncoderProcessor() *DataEncoderProcessor {
+	return &DataEncoderProcessor{
+		BaseMySQLDataProcessor{},
+	}
+}
+
+// ID return name of processor
+func (p *DataEncoderProcessor) ID() string {
+	return "DataEncoderProcessor"
+}
+
+// DataDecoderProcessor implements processor and encode/decode binary intX values to text format which acceptable by Tokenizer
+type DataDecoderProcessor struct {
+	BaseMySQLDataProcessor
+}
+
+// NewDataDecoderProcessor return new data encoder from/to binary format for tokenization
+func NewDataDecoderProcessor() *DataDecoderProcessor {
+	return &DataDecoderProcessor{
+		BaseMySQLDataProcessor{},
+	}
+}
+
+// ID return name of processor
+func (p *DataDecoderProcessor) ID() string {
+	return "DataDecoderProcessor"
+}
+
+// here we process encryption/tokenization results before send it to a client
+// acra decrypts or de-tokenize SQL literals, so we should convert string SQL literals to binary format
+// if client expects int, then parse INT literals and convert to binary 4/8 byte format
+// if expects bytes, then pass as is
+// if expects string, then leave as is if it is valid string or encode to hex
+// if it is encrypted data then we return default values or as is if applicable (binary data)
+func (p *BaseMySQLDataProcessor) encodeBinary(ctx context.Context, data []byte, setting config.ColumnEncryptionSetting, columnInfo base.ColumnInfo, logger *logrus.Entry) (context.Context, []byte, error) {
+	logger = logger.WithField("column", setting.ColumnName()).WithField("decrypted", base.IsDecryptedFromContext(ctx))
+	logger.Debugln("Encode binary")
+
+	if len(data) == 0 {
+		// we still need to encode result data as it might be null field in db
+		return ctx, PutLengthEncodedString(data), nil
+	}
+
+	var columnType = columnInfo.DataBinaryType()
+	dataTypeEncoded, isEncoded, err := p.encodeBinaryWithDataType(ctx, data, setting)
+	if err != nil && err != ErrConvertToDataType {
+		return nil, nil, err
+	}
+
+	// in case of successful encoding with defined data type return encoded data
+	if isEncoded {
+		return ctx, dataTypeEncoded, nil
+	}
+
+	// in case of error on converting to defined type we should roll back field type and encode it as it was originally
+	if err == ErrConvertToDataType {
+		ctx = base.MarkErrorConvertedDataTypeContext(ctx)
+		columnType = columnInfo.OriginBinaryType()
+	}
+
+	var encoded []byte
+	var isNumericType bool
+	// After processing, parse the value back and reencode it. Take care for the format to match.
+	// The result must have exact same format as it had. Overflows are unacceptable.
+	switch Type(columnType) {
+	case TypeNull:
+		if data != nil {
+			return nil, nil, errors.New("NULL not kept NULL")
+		}
+
+	case TypeTiny:
+		encoded = make([]byte, 1)
+		intValue, err := strconv.ParseInt(string(data), 10, 8)
```

Comment:
what about to use [BytesToString](https://github.com/cossacklabs/acra/blob/master/utils/utils.go#L236)?) 
looks like safe place for accessing data without concurrent access during parsing
same below for other int types

Reply:
Nice idea!